services/subscriber/service.go

```diff
@@ -0,0 +1,204 @@
+package subscriber
+
+import (
+	"expvar"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/meta"
+)
+
+// Statistics for the CQ service.
+const (
+	statPointsWritten = "points_written"
+)
+
+type PointsWriter interface {
+	WritePoints(p *cluster.WritePointsRequest) error
+}
+
+// unique set that identifies a given subscription
+type subEntry struct {
+	db   string
+	rp   string
+	name string
+}
+
+type Service struct {
+	subs      map[subEntry]PointsWriter
+	MetaStore interface {
+		Databases() ([]meta.DatabaseInfo, error)
+	}
+	NewPointsWriter func(u url.URL) (PointsWriter, error)
+	Logger          *log.Logger
+	statMap         *expvar.Map
+	ticker          *time.Ticker
+}
+
+func NewService() *Service {
+	return &Service{
+		subs:            make(map[subEntry]PointsWriter),
+		NewPointsWriter: newPointsWriter,
+		Logger:          log.New(os.Stderr, "[subscriber] ", log.LstdFlags),
+		statMap:         influxdb.NewStatistics("subscriber", "subscriber", nil),
+	}
+}
+
+func (s *Service) Open() error {
+	if s.MetaStore == nil {
+		panic("no meta store")
+	}
+	s.Update()
+	go s.pollForMetaUpdates()
+
+	return nil
+}
+
+func (s *Service) Close() error {
+	if s.ticker != nil {
+		s.ticker.Stop()
+	}
+	return nil
+}
+
+func (s *Service) pollForMetaUpdates() {
+	s.ticker = time.NewTicker(time.Minute)
+	for range s.ticker.C {
+		s.Update()
+	}
+}
+
+// start new and stop deleted subscriptions.
+func (s *Service) Update() error {
+	dbis, err := s.MetaStore.Databases()
+	if err != nil {
+		return err
+	}
+	allEntries := make(map[subEntry]bool, 0)
+	// Add in new subscriptions
+	for _, dbi := range dbis {
+		for _, rpi := range dbi.RetentionPolicies {
+			for _, si := range rpi.Subscriptions {
+				se := subEntry{
+					db:   dbi.Name,
+					rp:   rpi.Name,
+					name: si.Name,
+				}
+				allEntries[se] = true
+				if _, ok := s.subs[se]; ok {
+					continue
+				}
+				var bm BalanceMode
+				switch si.Mode {
+				case "ALL":
+					bm = ALL
+				case "ANY":
+					bm = ANY
+				default:
+					return fmt.Errorf("unknown balance mode %q", si.Mode)
+				}
+				writers := make([]PointsWriter, len(si.Destinations))
+				statMaps := make([]*expvar.Map, len(writers))
+				for i, dest := range si.Destinations {
+					u, err := url.Parse(dest)
+					if err != nil {
+						return err
+					}
+					w, err := s.NewPointsWriter(*u)
+					if err != nil {
+						return err
+					}
+					writers[i] = w
+					tags := map[string]string{"database": se.db, "retention_policy": se.rp, "name": se.name, "mode": si.Mode}
+					statMaps[i] = influxdb.NewStatistics(dest, "subscriber", tags)
```

Comment:
Absolutely!
I am going to have to get monitoring setup on my side of things too soon. That was really easy and nice to use.